### ISSUE TITLE
Join condition when expression is true fails to return records

### ISSUE THREAD

**Issue by chrisknoll** [duckdb/duckdb#3466]
#### What happens?
When performing a join between 2 tables using date columns, the join fails to return rows.   Excluding the join condition using the date columns works.

#### To Reproduce

This is the query without using the date columns in the join.  I've appended the join condition as a column in the select to confirm that the date comparison results in `true`:

```
select cohort_definition_id, subject_id, cohort_start_date, cohort_end_date, op1.observation_period_start_date, op1.observation_period_end_date,
  tc1.cohort_start_date >= op1.observation_period_start_date as gt_test,
  tc1.cohort_start_date <= op1.observation_period_end_date as lt_test
from main.cohort tc1
inner join main.observation_period op1 on tc1.subject_id = op1.person_id
--	and tc1.cohort_start_date >= op1.observation_period_start_date
--	and tc1.cohort_start_date <= op1.observation_period_end_date
where cohort_definition_id in (100);	
```

cohort_definition_id | subject_id | cohort_start_date | cohort_end_date | observation_period_start_date | observation_period_end_date | gt_test | lt_test
---------------------+------------+-------------------+-----------------+-------------------------------+-----------------------------+---------+--------
                 100 |          1 | 2002-12-25        | 2002-12-25      | 1963-12-31                    | 2010-01-01                  | true    | true   
                 100 |          1 | 2007-03-01        | 2007-03-01      | 1963-12-31                    | 2010-01-01                  | true    | true   
                 100 |          2 | 2003-03-01        | 2003-03-01      | 1963-12-31                    | 2010-01-01                  | true    | true   
                 100 |          2 | 2005-03-01        | 2005-03-01      | 1963-12-31                    | 2010-01-01                  | true    | true   

However, when uncommenting the two `and` clauses from the join, no records are returned:

```
select cohort_definition_id, subject_id, cohort_start_date, cohort_end_date, op1.observation_period_start_date, op1.observation_period_end_date,
  tc1.cohort_start_date >= op1.observation_period_start_date as gt_test,
  tc1.cohort_start_date <= op1.observation_period_end_date as lt_test
from main.cohort tc1
inner join main.observation_period op1 on tc1.subject_id = op1.person_id
	and tc1.cohort_start_date >= op1.observation_period_start_date
	and tc1.cohort_start_date <= op1.observation_period_end_date
where cohort_definition_id in (100);	
```

But the expression columns (gt test and lt test) both show TRUE as sthe result, so the problem is that the TRUE isn't leading to the successful join.

For clarity, this is the table definitions from information_schema (note the column types are properly defined as DATE):
table_name         | column_name                   | ordinal_position | data_type
-------------------+-------------------------------+------------------+----------
COHORT             | cohort_definition_id          |                1 | INTEGER  
COHORT             | subject_id                    |                2 | INTEGER  
COHORT             | cohort_start_date             |                3 | DATE     
COHORT             | cohort_end_date               |                4 | DATE     
OBSERVATION_PERIOD | observation_period_id         |                1 | INTEGER  
OBSERVATION_PERIOD | person_id                     |                2 | INTEGER  
OBSERVATION_PERIOD | observation_period_start_date |                3 | DATE     
OBSERVATION_PERIOD | observation_period_end_date   |                4 | DATE     
OBSERVATION_PERIOD | period_type_concept_id        |                5 | INTEGER  

#### Environment (please complete the following information):
 - OS: [Windows]
 - DuckDB Version: [0.3.2-2]
 - DuckDB Client: [JDBC]

**Comment by Mytherin** [comment]
Thanks for the report! Could you verify that this still occurs on the latest version (0.3.3)? If it does, could you also try to create a reproducible example?

**Comment by chrisknoll** [comment]
I used the [JDBC jar](https://github.com/duckdb/duckdb/releases/download/v0.3.3/duckdb_jdbc-windows-amd64.jar) on the 0.3.3 pre-release page, and I connected to an in-memory instance to run this test by using Sql Workbench/J with the JDBC url: jdbc:duckdb:

**Comment by hawkfish** [comment]
I can reproduce it in the test harness with the following:

```sql
statement ok
CREATE TABLE cohort (
	cohort_definition_id INTEGER,
	subject_id INTEGER,
	cohort_start_date DATE,
	cohort_end_date DATE,
);

statement ok
INSERT INTO cohort VALUES
	(100, 1, '2002-12-25', '2002-12-25'),
	(100, 1, '2007-03-01', '2007-03-01'),
	(100, 2, '2003-03-01', '2003-03-01'),
	(100, 2, '2005-03-01', '2005-03-01'),
;

statement ok
CREATE TABLE observation_period (
	 observation_period_id INTEGER,
	 person_id INTEGER,
	 observation_period_start_date DATE,
	 observation_period_end_date DATE,
	 period_type_concept_id INTEGER,
);

statement ok
INSERT INTO observation_period VALUES
	(1, 1, '1963-12-31', '2010-01-01', 1),
	(2, 2, '1963-12-31', '2010-01-01', 2),
;

query IIIIIIII
select cohort_definition_id, subject_id, cohort_start_date, cohort_end_date, op1.observation_period_start_date, op1.observation_period_end_date,
  tc1.cohort_start_date >= op1.observation_period_start_date as gt_test,
  tc1.cohort_start_date <= op1.observation_period_end_date as lt_test
from main.cohort tc1
inner join main.observation_period op1
        on tc1.subject_id = op1.person_id
       and tc1.cohort_start_date >= op1.observation_period_start_date
       and tc1.cohort_start_date <= op1.observation_period_end_date
where cohort_definition_id in (100);
----
100	1	2002-12-25	2002-12-25	1963-12-31	2010-01-01	true	true
100	1	2007-03-01	2007-03-01	1963-12-31	2010-01-01	true	true
100	2	2003-03-01	2003-03-01	1963-12-31	2010-01-01	true	true
100	2	2005-03-01	2005-03-01	1963-12-31	2010-01-01	true	true
```

**Comment by hawkfish** [comment]
The statistics propagation was eliminating join conditions that were always true, but not removing the corresponding statistics. Since all the conditions were 32 bit integers, no alarms went off, but we were using the underlying date range values to filter the id values. Hilarity ensued.

Simple 1 line fix, testing now.

**Comment by chrisknoll** [comment]
Thanks for putting together the test harness. I think I'm seeing other issues (which I'll file separate issues on them) but I'll include a complete 'test harness setup' script to demonstrate the issue next time.

Thanks for your analysis and fix!

**Comment by Mytherin** [comment]
Should be fixed by #3476
